firedrake/assemble.py

```diff
@@ -277,12 +381,68 @@ def get_matrix(expr, mat_type, sub_mat_type, *, bcs=None,
         raise ValueError("Monolithic matrix assembly not supported for systems "
                          "with R-space blocks")
 
-    tensor = matrix.Matrix(expr, bcs, mat_type, sparsity, ScalarType,
-                           options_prefix=options_prefix)
-    return tensor, (), lambda: tensor
+    return matrix.Matrix(expr, bcs, opts.mat_type, sparsity, ScalarType,
+                         options_prefix=opts.options_prefix)
+
+
+def _assemble_expr(expr, tensor, bcs, opts, assembly_rank):
+    """Assemble an expression.
+
+    :arg expr: The expression to be assembled.
+    :arg tensor: The tensor to write to.
+    :arg bcs: Iterable of boundary conditions.
+    :arg opts: :class:`_AssemblyOpts` containing the assembly options.
+    :arg assembly_rank: The appropriate :class:`_AssemblyRank`.
+    """
+    parloop_init_args = (expr, tensor, bcs, opts.diagonal, opts.fc_params, assembly_rank)
+    cached_init_args, cached_parloops = expr._cache.get("parloops", (None, None))
+    parloops = cached_parloops if cached_init_args == parloop_init_args else None
+
+    if not parloops:
+        parloops = _make_parloops(*parloop_init_args)
+        expr._cache["parloops"] = (parloop_init_args, parloops)
```

Comment:
Here, `(parloop_init_args, parloops)` overrides `(cached_init_args, cached_parloops)`, but don't we want to keep old values?

Reply:
This was intentional. I wanted to make sure that we didn't run into any data reference/memory usage issues. In particular with respect to the output tensor. The main use case that I've been targeting this refactor at is for repeated assembly of the same form (e.g. in a time loop) which will work fine.

Once parloops stop holding references to data (something I'm working on) I think this could be altered.

Reply:
Ok! Thanks.